// heres_your_route/dist/list.js
function list() {
    const string = `
    MIDDLEWARE verifyLogin() -> redirects login if body or params doesn't have user_id
    POST: api/signup
    POST: api/login
    POST: api/logout/:user_id
    POST: api/user/:user_id/problem
    POST: api/user/:user_id/cheatsheet
    POST: api/user/:user_id/task
    POST: api/user/:user_id/tag_snippet
    POST: api/user/:user_id/problem_task (needs problem_id, task_id in body)
    POST: api/user/:user_id/problem_cheatsheet (needs problem_id, cheatsheet_id in body)
    POST: api/user/:user_id/problem_tag_snippet (needs problem_id, tag_snippet_id in body)
    POST: api/user/:user_id/cheatsheet_task (needs cheatsheet_id, task_id in body)
    POST: api/user/:user_id/task_tag_snippet (needs task_id, tag_snippet_id in body)
    POST: api/user/:user_id/cheatsheet_tag_snippet (needs cheatsheet_id, tag_snippet_id in body)
    PUT: api/user/:user_id/problem -> needs problem_id in body
    PUT: api/user/:user_id/cheatsheet -> needs cheatsheet_id in body
    PUT: api/user/:user_id/task -> needs task_id in body
    PUT: api/user/:user_id/tag_snippet -> needs tag_snippet_id in body 
    DELETE: api/user/:user_id/problem/:id 
    DELETE: api/user/:user_id/cheatsheet/:id 
    DELETE: api/user/:user_id/task/:id 
    DELETE: api/user/:user_id/tag_snippet/:id 
    GET: api/user/:user_id/problem
    GET: api/user/:user_id/cheatsheet
    GET: api/user/:user_id/task
    GET: api/user/:user_id/tag_snippet
    GET: api/user/:user_id/problem/:id 
    GET: api/user/:user_id/cheatsheet/:id 
    GET: api/user/:user_id/task/:id 
    GET: api/user/:user_id/tag_snippet/:id 
    GET(SEARCH): api/user/:user_id/search/problem/:query
    GET(SEARCH): api/user/:user_id/search/cheatsheet/:query
    GET(SEARCH): api/user/:user_id/search/task/:query
    GET(SEARCH): api/user/:user_id/search/tag_snippet/:query
    GET(SORTS): api/user/:user_id/created/:table
    GET(SORTS): api/user/:user_id/important/:table
    GET(SORTS): api/user/:user_id/end/:table
    GET(SORTS): api/user/:user_id/edit/:table
    GET(SORTS): api/user/:user_id/difficult/:table
    GET(SORTS): api/user/:user_id/public/:table
    GET(SORTS): api/user/:user_id/seconds/:table
    GET(SORTS): api/user/:user_id/usefulness/:table
    GET(SORTS): api/user/:user_id/joy/:table
    PAGES: /signup
    PAGES: /login
    PAGES: /forgot_pass 
    PAGES: /dashboard/:user_id  (get all problem joins)
    PAGES: /calendar/:user_id  (get all task joins)
    PAGES: /table/:user_id  (get all problem joins)
    PAGES: /profile/:user_id  (get all problem joins)
    PAGES: /settings/:user_id (get all problem joins)

  `
    console.log(string)
    return string;
}



module.exports = { list };
