// lesleys_hbs/dist/date.js
const when_you_free = require("when_you_free")


class HelpersDate {
    /**
     * let helpersDate = new HelpersDate()
     * @returns {any}
     */
    constructor() {

    }

    /**
     * @example
     * {{getDuration "00:20:00" "00:40:00" "minutes"}}
     */
    getDuration(dateOne, dateTwo, type) {
            return when_you_free.getDuration(dateOne, dateTwo, type)
        }
        /**
         * @example
         * {{getReadableFormat problem_created from/to}} // 15 hours ago 
         */
    getReadableFormat(date, type) {
            return when_you_free.getReadableFormat(date, type)
        }
        /**
         * @example {{getDifferenceInDays created "days"}}
         * @date 2022-02-17
         */
    difference(one, two, type) {
            return when_you_free.difference(one, two, type)
        }
        /**
         * @example
         * {{format created "ll"}}
         * @date 2022-02-17
         */
    format(date, type) {
        return when_you_free.format(date, type)
    }
    dateIsWithinLimit(date, days) {
        return when_you_free.dateIsWithinLimit(date, days)
    }
    isBefore(date, date2 = undefined) {
        if (date2 !== undefined) {
            return when_you_free.isBefore(date, date2)
        } else {
            return when_you_free.isBefore(date, new Date())
        }
    }
    formatDateToPost(date) {
        return when_you_free.formatDateToPost(date)
    }
    formatToDateInput(date) {
        // yyyy-mm-ll
        return when_you_free.format(date, "YYYY-MM-DD")
    }

}
let helpersDate = new HelpersDate()

module.exports = {
    helpersDate
}
