// backend_for_noobs/dist/delete.js
const ef = require("effective_knex")
const { getOneProblemById, getOneCheatsheetById, getOneTaskById, getOneTagSnippetById } = require("./getOne.js")
const { knex } = require("./knex")

async function deleteUser(knex, user_id) {
    let del = await ef.delByKeyValue(knex, "tag_snippet", "user_id", user_id)
    let del2 = await ef.delByKeyValue(knex, "task", "user_id", user_id)
    let del3 = await ef.delByKeyValue(knex, "cheatsheet", "user_id", user_id)
    let del4 = await ef.delByKeyValue(knex, "problem", "user_id", user_id)
    let del5 = await ef.delByKeyValue(knex, "hourglass", "user_id", user_id)
    let del6 = await ef.delByKeyValue(knex, "user", "id", user_id)
    return del6;
}
async function deleteProblem(knex, user_id, problem_id) {
    let hourglass = await getOneProblemById(knex, user_id, problem_id)
    let hourglass_id = hourglass.hourglass_id
    let del = await ef.delByKeyValue(knex, "tag_snippet", "problem_id", problem_id)
    let del2 = await ef.delByKeyValue(knex, "task", "problem_id", problem_id)
    let del3 = await ef.delByKeyValue(knex, "cheatsheet", "problem_id", problem_id)
    let del4 = await ef.delByKeyValue(knex, "problem", "id", problem_id)
    let del5 = await ef.delByKeyValue(knex, "hourglass", "id", hourglass_id)
    return del5;
}
async function deleteCheatsheet(knex, user_id, cheatsheet_id) {
    let hourglass = await getOneCheatsheetById(knex, user_id, cheatsheet_id)
    let hourglass_id = hourglass.hourglass_id
    let del = await ef.delByKeyValue(knex, "tag_snippet", "cheatsheet_id", cheatsheet_id)
    let del2 = await ef.delByKeyValue(knex, "task", "cheatsheet_id", cheatsheet_id)
    let del3 = await ef.delByKeyValue(knex, "cheatsheet", "id", cheatsheet_id)
    let del5 = await ef.delByKeyValue(knex, "hourglass", "id", hourglass_id)
    return del5;
}
async function deleteTask(knex, user_id, task_id) {
    let hourglass = await getOneTaskById(knex, user_id, task_id)
    let hourglass_id = hourglass.hourglass_id
    let del = await ef.delByKeyValue(knex, "tag_snippet", "task_id", task_id)
    let del3 = await ef.delByKeyValue(knex, "task", "id", task_id)
    let del5 = await ef.delByKeyValue(knex, "hourglass", "id", hourglass_id)
    return del5;
}
async function deleteTagSnippet(knex, user_id, tag_snippet_id) {
    let hourglass = await getOneTagSnippetById(knex, user_id, tag_snippet_id)
    let hourglass_id = hourglass.hourglass_id
    let del2 = await ef.delByKeyValue(knex, "tag_snippet", "id", tag_snippet_id)
    let del5 = await ef.delByKeyValue(knex, "hourglass", "id", hourglass_id)
    return del5;
}

async function testDelete() {

    let sampleUser = {
        email: "ddasdffasdfasdfdsf@bu.edu",
        password: "testtest"
    }

    let sampleObject = {
        user_id: 1,
        problem: "we are scared",
        whatshouldbe: "we should be less frightened",
        whatactuallyis: "we get scared all the time",
        hypothesis: "don't forget who loved you",
        plan: "Depend on love, rather than validation",
        difficulty: 2,
        usefulness: 3,
        end: "2022-03-27"
    }
    let post_problem = await postProblem(knex, sampleObject)
    let sampleCheatsheet = {
        user_id: 1,
        problem_id: post_problem.problem_id,
        importance: 2,
        usefulness: 3,
        title: "getting things done",
        key_info: "just do it",
        description: "be more like water",
        structure: "think less"
    }
    let post_cheatsheet = await postCheatsheet(knex, sampleCheatsheet)
    let sampleTask = {
        user_id: 1,
        problem_id: post_problem.problem_id,
        cheatsheet_id: post_cheatsheet.cheatsheet_id,
        name: "KEEP IT SIMPLE STUPID",
        difficulty: 2,
        usefulness: 1,
        importance: 1,
        end: "2022-03-15"
    }
    let post_task = await postTask(knex, sampleTask)

    let sampleTagSnippet = {
        task_id: post_task.task_id,
        problem_id: post_problem.problem_id,
        cheatsheet_id: post_cheatsheet.cheatsheet_id,
        tag_snippet_id: 18,
        hourglass_id: 330,
        tag: 'quote',
        snippet: 'be more selfish with your time. this is your life. reach your own goals. be more confident in yourself, you got this. i want cyrus on my team',
        user_id: 1,
        importance: 1,
        difficulty: 2,
        public: true,
        status: 'to do',
        usefulness: 1,
    }
    let post_tag = await postTagSnippet(knex, sampleTagSnippet)
        // let first_delete = await deleteProblem(knex, 1, post_problem.problem_id)
        // console.log("🚀 ~ file: delete.js ~ line 127 ~ testDelete ~ first_delete", first_delete)
        // console.log(first_delete, "DELETED")
        // let second_delete = await deleteCheatsheet(knex, 1, post_cheatsheet.cheatsheet_id)
        // console.log("🚀 ~ file: delete.js ~ line 130 ~ testDelete ~ second_delete", second_delete)
        // console.log("🚀 ~ file: delete.js ~ line 129 ~ testDelete ~ second_delete", second_delete)
        //     // console.log("🚀 ~ file: delete.js ~ line 129 ~ testDelete ~ second_delete", second_delete)

    // let third_delete = await deleteTask(knex, 1, post_task.task_id)
    // console.log("🚀 ~ file: delete.js ~ line 132 ~ testDelete ~ third_delete", third_delete)
    //     // console.log("🚀 ~ file: delete.js ~ line 129 ~ testDelete ~ second_delete", third_delete)
    let fourth_delete = await deleteTagSnippet(knex, 1, post_tag.tag_snippet_id)

    console.log(fourth_delete)
}
// testDelete()

module.exports = { deleteUser, deleteProblem, deleteCheatsheet, deleteTask, deleteTagSnippet, testDelete };
