// backend_for_noobs/dist/follow.js
const { knex } = require("./knex")
const { upset } = require("whats_wrong")
const ef = require("effective_knex")
const ROUTE = "/api/user/:user_id/"
async function postFollow(knex, followObject) {
    try {
        let postHourglass = await ef.post(knex, "hourglass", followObject)
        followObject["hourglass_id"] = postHourglass.id
        let followed = await ef.post(knex, "follow", followObject)
        return followed
    } catch (error) {
        return upset(error, "post follow", "should be able to post user")
    }
}

async function unfollow(followObject) {
    try {
        let getOne = await ef.getByObject(knex, "follow", { user_id: followObject.user_id, follow_user_id: followObject.follow_user_id })
        let finalGetOne = ef.getObject(getOne)
        let remove = await ef.del(knex, "follow", finalGetOne.id)
    } catch (error) {
        return upset(error, "unfollow function", "should be able to unfollow user")
    }
}

async function testFollow() {


    let sampleUser = {
        email: "lesleyc@bu.edu",
        password: "testtest"
    }
    let followThisUser = {
        email: "ryan@bu.edu",
        password: "testtest"
    }


}

module.exports = { postFollow, unfollow, testFollow };
