// backend_for_noobs/dist/getOne.js
const arrz = require("array_helperz")
const { getAllProblems, getAllCheatsheets, getAllTasks, getAllTagSnippets, } = require("./getAll.js")
const ef = require("effective_knex")
const { knex } = require("./knex")

async function getOneProblemById(knex, user_id, problem_id) {
    let result;
    let problems = await getAllProblems(knex, user_id)
    if (problems.length > 0) {
        result = arrz.filterWithFuncObjectOrString(problems, { user_id: user_id, problem_id: problem_id })

        let object = ef.getObject(result)
        object["table_name"] = "problem"
        return object;
    } else {
        return []
    }
}
async function getOneCheatsheetById(knex, user_id, cheatsheet_id) {
    let cheatsheets = await getAllCheatsheets(knex, user_id)

    let result;
    if (cheatsheets.length > 0) {
        result = arrz.filterWithFuncObjectOrString(cheatsheets, { user_id: user_id, cheatsheet_id: cheatsheet_id })
        let object = ef.getObject(result)
        object["table_name"] = "cheatsheet"
        return object;
    } else {
        return []
    }

}

async function getOneTaskById(knex, user_id, task_id) {
    let tasks = await getAllTasks(knex, user_id)
    let result;
    if (tasks.length > 0) {
        result = arrz.filterWithFuncObjectOrString(tasks, { user_id: user_id, task_id: task_id })
        let object = ef.getObject(result)
        object["table_name"] = "task"
        return object;
    } else {
        return []
    }
}
async function getOneTagSnippetById(knex, user_id, tag_snippet_id) {
    let tagSnippets = await getAllTagSnippets(knex, user_id)

    let result;
    if (tagSnippets.length > 0) {
        result = arrz.filterWithFuncObjectOrString(tagSnippets, { user_id: user_id, tag_snippet_id: tag_snippet_id })
        let object = ef.getObject(result)
        object["table_name"] = "tag_snippet"
        return object;
    } else {
        return []
    }
}

async function testGetOne() {
    let problem = await getOneProblemById(knex, 1, 1)
    console.log("🚀 ~ file: getOne.js ~ line 66 ~ testGetOne ~ problem", problem)
    let cheatsheet = await getOneCheatsheetById(knex, 1, 3)
    let task = await getOneTaskById(knex, 1, 79)
    let tag = await getOneTagSnippetById(knex, 1, 18)

}

// testGetOne()

module.exports = { getOneProblemById, getOneCheatsheetById, getOneTaskById, getOneTagSnippetById, testGetOne };
