// backend_for_noobs/dist/oneToManyPost.js
const ef = require("effective_knex")
const { knex } = require("./knex")
const { getOneProblemCheatsheets, getOneProblemTasks, getOneProblemTagSnippets, getOneCheatsheetTask, getOneCheatsheetTagSnippet, getOneTaskTagSnippets, } = require("./oneToMany")

async function addCheatsheetToProblem(knex, user_id, problem_id, cheatsheet_id) {
    let updateCheatsheet = await ef.update(knex, "cheatsheet", cheatsheet_id, { problem_id: problem_id })
    let getProblem = await getOneProblemCheatsheets(knex, user_id, problem_id)
    return getProblem;
}



async function addTaskToProblem(knex, user_id, problem_id, task_id) {
    let updateTask = await ef.update(knex, "task", task_id, { problem_id: problem_id })
    let getProblem = await getOneProblemTasks(knex, user_id, problem_id)
    return getProblem;
}
async function addTagSnippetToProblem(knex, user_id, problem_id, tag_snippet_id) {
    let updateTag = await ef.update(knex, "tag_snippet", tag_snippet_id, { problem_id: problem_id })
    let getProblem = await getOneProblemTagSnippets(knex, user_id, problem_id)
    return getProblem;
}


async function addTaskToCheatsheet(knex, user_id, cheatsheet_id, task_id) {
    let updateTask = await ef.update(knex, "task", task_id, { cheatsheet_id: cheatsheet_id })
    let getCheatsheet = await getOneCheatsheetTask(knex, user_id, cheatsheet_id)
    return getCheatsheet;
}


async function addTagSnippetToCheatsheet(knex, user_id, cheatsheet_id, tag_snippet_id) {
    let updateTag = await ef.update(knex, "tag_snippet", cheatsheet_id, { cheatsheet_id: cheatsheet_id })
    let getCheatsheet = await getOneCheatsheetTagSnippet(knex, user_id, cheatsheet_id)
    return getCheatsheet;
}


async function addTagSnippetToTask(knex, user_id, task_id, tag_snippet_id) {
    let updateTag = await ef.update(knex, "tag_snippet", task_id, { task_id: task_id })
    let getTask = await getOneTaskTagSnippets(knex, user_id, task_id)
    return getTask
}


async function testOneToManyPost() {
    let add_cheatsheet_to_problem = await addCheatsheetToProblem(knex, 1, 139, 10)
    let add_task_to_problem = await addTaskToProblem(knex, 1, 139, 4)
    let add_tag_to_problem = await addTagSnippetToProblem(knex, 1, 139, 1)
    let add_task_to_cheatsheet = await addTaskToCheatsheet(knex, 1, 10, 4)
    let add_tag_to_cheatsheet = await addTagSnippetToCheatsheet(knex, 1, 10, 1)
    let add_tag_to_task = await addTagSnippetToTask(knex, 1, 10, 1)
}
// testOneToManyPost()

module.exports = { addCheatsheetToProblem, addTaskToProblem, addTagSnippetToProblem, addTaskToCheatsheet, addTagSnippetToCheatsheet, addTagSnippetToTask, testOneToManyPost };
