// backend_for_noobs/dist/public.js
const _ = require("lodash")
const arrz = require("array_helperz")
const { getAllTagSnippetJoins, getAllProblemJoins, getAllCheatsheetJoins, getAllTaskJoins } = require("./oneToMany.js")
const make_legit = require("make_legit")
const { knex } = require("./knex")


async function getAllPublicProblems(knex, user_id) {
    let problems = await getAllProblemJoins(knex, user_id)
    let filtered = _.filter(problems, { public: true })
    return filtered;
}

async function getAllPublicCheatsheets(knex, user_id) {
    let problems = await getAllCheatsheetJoins(knex, user_id)
    let filtered = _.filter(problems, { public: true })
    return filtered;
}

async function getAllPublicTasks(knex, user_id) {
    let problems = await getAllTaskJoins(knex, user_id)
    let filtered = _.filter(problems, { public: true })
    return filtered;
}
async function getAllPublicTagSnippets(knex, user_id) {
    let problems = await getAllTagSnippetJoins(knex, user_id)
    let filtered = _.filter(problems, { public: true })
    return filtered;
}
async function publicToPrivateProblems(knex, user_id) {
    let merged = await getAllProblemJoins(knex, user_id)
    let sorted = arrz.sortDesc(merged, "public", "boolean")
    return sorted
}

async function publicToPrivateCheatsheets(knex, user_id) {
    let trimmed = await getAllCheatsheetJoins(knex, user_id)
    let sorted = arrz.sortDesc(trimmed, "public", "boolean")
    return sorted
}
async function publicToPrivateTasks(knex, user_id) {
    let trimmed = await getAllTaskJoins(knex, user_id)
    let sorted = arrz.sortDesc(trimmed, "public", "boolean")
    return sorted
}
async function publicToPrivateTagSnippets(knex, user_id) {
    let trimmed = await getAllTagSnippetJoins(knex, user_id)
    let sorted = arrz.sortDesc(trimmed, "public", "boolean")
    return sorted
}
async function testpublicToPrivate() {
    let problems = await publicToPrivateProblems(knex, 1)
    console.log(problems[0])
    console.log(problems[problems.length - 1])
    let tasks = await publicToPrivateTasks(knex, 1)
    let tagsnippets = await publicToPrivateTagSnippets(knex, 1)
    let cheatsheets = await publicToPrivateCheatsheets(knex, 1)
    let publicProblems = await getAllPublicProblems(knex, 1)
    console.log("🚀 ~ file: public.js ~ line 66 ~ testpublicToPrivate ~ publicProblems", publicProblems[0])
    let publicCheatsheets = await getAllPublicCheatsheets(knex, 1)
    let publicTasks = await getAllPublicTasks(knex, 1)
    let publicTags = await getAllPublicTagSnippets(knex, 1)

}

module.exports = { getAllPublicProblems, getAllPublicCheatsheets, getAllPublicTasks, getAllPublicTagSnippets, publicToPrivateProblems, publicToPrivateCheatsheets, publicToPrivateTasks, publicToPrivateTagSnippets, testpublicToPrivate };
